src/choice_map.jl

```diff
@@ -953,3 +953,24 @@ nested_view(c::ChoiceMap) = ChoiceMapNestedView(c)
 # aux data together.
 
 export nested_view
+
+#############################################
+# filtering choice maps based on selections #
+#############################################
+
+function get_selected(
+        choices::ChoiceMap, selection::Selection)
+    output = choicemap()
+    for (key, value) in get_values_shallow(choices)
+        if (key in selection)
+            output[key] = value
+        end
+    end
+    for (key, submap) in get_submaps_shallow(choices)
+        subselection = selection[key]
+        set_submap!(output, key, get_selected(submap, subselection))
+    end
+    output
+end
```

Comment:
Can we add `iter_deep` from https://github.com/probcomp/Gen/issues/123 and then have this function be something like

```julia
get_selected(choices::ChoiceMap, selection::Selection) = choicemap(
    ((keypath, choices[keypath]) for keypath in iter_deep(choices)
    if matches(keypath, selection))...)
```
?

Reply:
I agree that `iter_deep` is useful for certain use cases, but my concern with using it here (and elsewhere in the internal implementations of the choice maps API) is that `matches(keypath, selection)` is linear in the depth of the address. Simultaneously doing a walk of two trees is a common implementation technique in the Gen implementation that avoids this.

Reply:
Is that still true if we require `matches` to depend only on the bottom
level key (as appears to be already the case in the existing implementation)?

The principle is that I want to separate the traversal logic from the matching logic, if feasible.  So I'm trying to understand if/why the separation is not possible in this case.

Reply:
Oops, I just reread the code and saw that there's a traversal of Klkeypaths in the selection too.  OK, thinking about it again.

Reply:
Seems reasonable, up to me not knowing the motivation for having the current implementations of choice maps use the data structures it uses, instead of dl nested dicts.  Let's call this resolved now, and I'll ask you later for some info on the above.